docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,286 @@
+    checksums:
+      sha256: d16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c
+```
+
+#### resolving the download url
+```python
+# parse the purl
+pkg:huggingface/{namespace}/{name}@{commit_hash}[?repository_url={repo_url}][&file_name={file_name}]
+
+# download URL (in case a file_name is present)
+https://{repository_url}/{namespace}/{name}/blob/{commit_hash}/{file_name}
+
+# git clone command (in case file_name is absent) 
+git clone https://{repository_url}/{namespace}/{name}
+git checkout {commit_hash}
+```
+
+#### SBOM component
+
+```json
+[
+    {
+        "name": "Llama-3.1-Nemotron-70B-Instruct-HF",
+        "purl": "pkg:huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF@043235d6088ecd3dd5fb5ca3592b6913fd516027",
+        "properties": [
+            {
+                "name": "cachi2:found_by",
+                "value": "cachi2:generic"
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF"
+            }
+        ],
+        "type": "library"
+    },
+    {
+        "name": "Llama-3.1-Nemotron-70B-Instruct-HF",
+        "purl": "pkg:huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF@043235d6088ecd3dd5fb5ca3592b6913fd516027?file_name=model-00001-of-00030.safetensors&checksum=sha256%3Ad16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c",
+        "properties": [
+            {
+                "name": "cachi2:found_by",
+                "value": "cachi2:generic"
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/fac73d3507320ec1258620423469b4b38f88df6e/model-00001-of-00030.safetensors"
+            }
+        ],
+        "type": "file"
+    }
+]
+```
+
+### OCI
+
+#### generic_artifacts.yaml
+```yamlit
+#### resolving the download url\
+```python
+# parse the input purl
+pkg:oci/{name}@{algorithm}%3A{digest}?repository_url={repository_url_and_namespace}[&arch={arch}][&tag={tag}]
+
+# generate the download command
+podman pull {repository_url_and_namespace}/{name}@{algorithm}:{digest}
+```
+
+#### SBOM component
+```json
+{
+    "name": "buildah-task",
+    "purl": "pkg:oci/buildah-task@sha256%3Ab2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c?repository_url=quay.io/konflux-ci&arch=amd64&tag=latest",
+    "properties": [
+        {
+            "name": "cachi2:found_by",
+            "value": "cachi2:generic"
+        }
+    ],
+    "externalReferences": [
+        {
+            "type": "distribution",
+            "url": "quay.io/konflux-ci/buildah-task@sha256:b2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c"
+        }
+    ],
+    "type": "container"
+}
+```
+
+## Alternative: consume options instead of a purl
```

Comment:
@aloubyansky How do you feel about this alternate proposal? Would it be difficult to generate the lockfiles for PNC/Maven artifacts?

My initial feeling is that having the attributes broken down is clearer and avoids the need to parse the purl according to the spec.  The flexibilty to define the output purl would still exist, since we can add any needed attributes in order to generate the purl we want.

Reply:
I haven't read the alternative yet, but based on your comment, it is effectively the effect your comment above: https://github.com/containerbuildsystem/cachi2/pull/692#discussion_r1828416919?

It is a recognition that different purl types will have different field requirements, so you would expose _those_ fields as being required?

It seems like it would be harder to validate the validity of fields because it could be possible for a user to add content to an attribute which would end up matching some other part of the purl spec (i.e. if control characters are not properly escaped). If you default to consuming a purl itself, then you could validate the parsing of the purl and then only update those specific sub-elements which are relevant for cachi2 to control (i.e. uri/hash).

Reply:
To record some parts of a conversation that I had out-of-band:

* I don't see an issue with this approach, but you should ensure that you are making the decision for the proper reason.
* With this method, I think it would still be required to validate the entire purl after it is assembled. Therefore, I don't think that this would alleviate the need to parse a purl
* This seems to mostly affect user input. Is it a better user experience to provide values based on some set of required/optional keys? Would users be more likely to do this properly than if they were to specify the purl in its entirety? 
* Will the verbosity (i.e. the increased number of lines in the configuration file) be an issue? Or would it be a desired feature?